packages/backend/server/src/plugins/copilot/providers/perplexity.ts

```diff
@@ -0,0 +1,209 @@
+import assert from 'node:assert';
+
+import { EventSourceParserStream } from 'eventsource-parser/stream';
+
+import {
+  CopilotPromptInvalid,
+  CopilotProviderSideError,
+  metrics,
+} from '../../../base';
+import {
+  CopilotCapability,
+  CopilotChatOptions,
+  CopilotProviderType,
+  CopilotTextToTextProvider,
+  PromptMessage,
+} from '../types';
+
+export type PerplexityConfig = {
+  apiKey: string;
+};
+
+interface Message {
+  role: 'assistant';
+  content: string;
+}
+
+interface Choice {
+  message: Message;
+  delta: Message;
+  finish_reason: 'stop' | null;
+}
+
+interface PerplexityData {
+  citations: string[];
+  choices: Choice[];
+}
+
+export function injectCitations(content: string, citations: string[]) {
+  // Match [[n]] and [n] patterns
+  // Not match if they're already part of a formatted citation
+  const regex = /(?:\[\[(\d+)\]\]|\[(\d+)\])(?!\]?\([^)]*\))/g;
```

Comment:
well, I don't like regex at all, I prefer hand writing parser. But let's keep it until bugs occurs.

Reply:
Well, I added a "CitationParser" class for citation parsing, supplemented by related tests.